code/pilotfile/csg.cpp

```diff
@@ -1018,40 +1018,36 @@ void pilotfile::csg_read_variables()
 {
 	int idx;
 
-	Campaign.num_variables = cfread_int(cfp);
+	int num_variables = cfread_int(cfp);
 
-	if (Campaign.num_variables > 0) {
-		Campaign.variables = (sexp_variable *) vm_malloc( Campaign.num_variables * sizeof(sexp_variable) );
-		Verify( Campaign.variables != NULL );
+	if (num_variables > 0) {
 
-		memset( Campaign.variables, 0, Campaign.num_variables * sizeof(sexp_variable) );
-
-		for (idx = 0; idx < Campaign.num_variables; idx++) {
-			Campaign.variables[idx].type = cfread_int(cfp);
-			cfread_string_len(Campaign.variables[idx].text, TOKEN_LENGTH, cfp);
-			cfread_string_len(Campaign.variables[idx].variable_name, TOKEN_LENGTH, cfp);
+		for (idx = 0; idx < num_variables; idx++) {
+			sexp_variable temp_var;
+			memset(&temp_var, 0, sizeof(sexp_variable));
+			temp_var.type = cfread_int(cfp);
+			cfread_string_len(temp_var.text, TOKEN_LENGTH, cfp);
+			cfread_string_len(temp_var.variable_name, TOKEN_LENGTH, cfp);
+			Campaign.persistent_variables.push_back(temp_var);
 		}
 	}
 
 	if (csg_ver < 4) { // CSG files before version 4 don't have a Red Alert set of CPVs to load, so just copy the regular set.
-		Campaign.redalert_num_variables = Campaign.num_variables;
-		Campaign.redalert_variables = (sexp_variable *) vm_malloc( Campaign.redalert_num_variables * sizeof(sexp_variable) );
-		Verify( Campaign.redalert_variables != NULL);
-
-		memcpy( Campaign.redalert_variables, Campaign.variables, Campaign.num_variables * sizeof(sexp_variable));
+		Campaign.persistent_variables.clear();
+		for (int i = 0; i < Campaign.persistent_variables.size(); i++) {
+			Campaign.red_alert_data.push_back(Campaign.persistent_variables[i]);
+		}
 	} else {
-		Campaign.redalert_num_variables = cfread_int(cfp);
-
-		if (Campaign.redalert_num_variables > 0) {
-			Campaign.redalert_variables = (sexp_variable *) vm_malloc( Campaign.redalert_num_variables * sizeof(sexp_variable) );
-			Verify( Campaign.redalert_variables != NULL );
-
-			memset( Campaign.redalert_variables, 0, Campaign.redalert_num_variables * sizeof(sexp_variable) );
-
-			for (idx = 0; idx < Campaign.redalert_num_variables; idx++) {
-				Campaign.redalert_variables[idx].type = cfread_int(cfp);
-				cfread_string_len(Campaign.redalert_variables[idx].text, TOKEN_LENGTH, cfp);
-				cfread_string_len(Campaign.redalert_variables[idx].variable_name, TOKEN_LENGTH, cfp);
+		int redalert_num_variables = cfread_int(cfp);
+
+		if (redalert_num_variables > 0) {
```

Comment:
I'm not sure if it's necessary but `red_alert_data` should probably be clear somewhere around here to remove any old data.